Add pref to allow indenting line comments

marcelgerber opened the pull request:
Fixes #8012.
This pref is enabled by default, aligned with https://github.com/adobe/brackets/issues/8012#issuecomment-54155264.

The pref does not apply when there's no line comment syntax available and we fall back to block commenting the line.

cc @le717 @redmunds

Commit:
Add pref to allow indenting line comments

sprintr commented on src/editor/EditorCommandHandlers.js:
```diff
@@ -202,18 +207,26 @@ define(function (require, exports, module) {
             // Comment out - prepend the first prefix to each line
             for (i = startLine; i <= endLine; i++) {
-                editGroup.push({text: prefixes[0], start: {line: i, ch: 0}});
+                if (PreferencesManager.get("indentLineComment")) {
```
This pref is defined as `"lineComment.indent"` above.

MarcelGerber replied:
Thanks. I renamed the pref just before pushing and forgot about this occurence.
Fixed now.

Commit:
Use correct pref name

redmunds commented:
@MarcelGerber Making this a preference (which can be specified per-language) is a good idea for languages where indenting on single-line comments is a user choice.

Can you also add a new property in languages.json file for languages that can *only* work as indented (or not) ? Otherwise, languages such as Jade won't work correctly when Brackets is initially installed -- users will need to manually set this preference first (and need to *know* to do this). There are some other languages (vbscript? bash?) that only work with single-line comments at start of line (so we can't simply change default to indented).

I think setting in languages.json should be the default, and preference will override it.

Please add a unit test.

TomMalbran commented:
This is great. I have been trying to implement this, but I had some issues with other changes I made and didn't had much time to fix them. So I don't mind that you implemented it.

But when I implemented it, I finished with 3 modes:
1. If the language had indented comments, the comment would be added right after the leading space. I did added this in language.json and could not be overwritten by other preferences, since commenting might not work.

2. I then added a preference to switch to indenting comments, this would add the line comments right after the minimum leading space in the selected block, so that the line prefixes where all the same indentation.

3. If that preference was set to false, then it would use the current commenting at the beginning of each line.


I would suggest something like this. If comments on a language only work when they are right before the content, then it shouldn't be possible to change this behaviour with a preference. If there are languages where the line comment needs to be at the start of the line, then we would need another attribute in language.json for those languages which should not be overwritten by a preference.

MarcelGerber commented:
Ah, I found your work in progress at https://github.com/adobe/brackets/commit/1a87b7e56fca48cb2902037f22e5ad2ef2d82c22#diff-a4c3174140581e8c106b11103021b6a8. I might be able to reuse some of that. Thanks for pointing out!

Commit:
Let languages override the comment indentation settings

MarcelGerber commented:
Ready for another review.
I've implemented all 3 modes and implemented the possibility for a language to force comments at the beginning of line.

Unit tests still missing out, though.

I've tested `.bat` and `.sh` and it looks like both don't explicitly need to have the comment at the beginning (couldn't test VBScript). Also, SASS compiled just fine with comments not correctly indented.
That's why I only added Stylus for now.

petetnt commented:
#11954 landed which superceded this PR

MarcelGerber commented:
Closing.
